experimental/skills/DialogToDialog/DialogRootBot/Dialogs/SkillDialog.cs

```diff
@@ -0,0 +1,153 @@
+            }
+
+            _botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;
+            if (string.IsNullOrWhiteSpace(_botId))
+            {
+                throw new ArgumentException($"{MicrosoftAppCredentials.MicrosoftAppIdKey} is not in configuration");
+            }
+
+            _skillClient = skillClient ?? throw new ArgumentNullException(nameof(skillClient));
+            _skillsConfig = skillsConfig ?? throw new ArgumentNullException(nameof(skillsConfig));
+            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
+            _activeSkillProperty = conversationState.CreateProperty<string>($"{typeof(SkillDialog).FullName}.ActiveSkillProperty");
+        }
+
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default)
+        {
+            if (options is CancellationToken)
+            {
+                throw new ArgumentException($"{nameof(options)} cannot be a cancellation token");
+            }
+
+            if (!(options is SkillDialogArgs dialogArgs))
+            {
+                throw new ArgumentNullException(nameof(options), $"Unable to cast {nameof(options)} to {nameof(SkillDialogArgs)}");
+            }
+
+            var skillId = dialogArgs.SkillId;
+            if (!_skillsConfig.Skills.TryGetValue(skillId, out var skillInfo))
+            {
+                throw new KeyNotFoundException($"Unable to find \"{skillId}\" in the skill configuration.");
+            }
+
+            // Store Skill ID for this dialog instance
+            await _activeSkillProperty.SetAsync(dc.Context, skillId, cancellationToken);
+
+            await dc.Context.SendTraceActivityAsync($"{GetType().Name}.BeginDialogAsync(). Using activity of type: {dialogArgs.ActivityType}", cancellationToken: cancellationToken);
+
+            Activity skillActivity;
+            switch (dialogArgs.ActivityType)
+            {
+                case ActivityTypes.Event:
+                    var eventActivity = Activity.CreateEventActivity();
+                    eventActivity.Name = dialogArgs.Name;
+                    eventActivity.ApplyConversationReference(dc.Context.Activity.GetConversationReference(), true);
+                    skillActivity = (Activity)eventActivity;
+                    break;
+
+                case ActivityTypes.Message:
+                    var messageActivity = Activity.CreateMessageActivity();
+                    messageActivity.Text = dialogArgs.Text;
+                    skillActivity = (Activity)messageActivity;
+                    break;
+
+                default:
+                    throw new ArgumentException($"Invalid activity type in {dialogArgs.ActivityType} in {nameof(SkillDialogArgs)}");
+            }
+
+            ApplyParentActivityProperties(dc, skillActivity, dialogArgs);
```

Comment:
Is there a reason not to make ApplyParentActivityProperties part of SendToSkill?